File: js/src/engine/sound-manager.js
'use strict';

import Sound from './sound';

const [SOURCES, AUDIOCONTEXT, PURGE, LOAD, CURRENT_MUSIC, REFERENCES] = [Symbol(), Symbol(), Symbol(), Symbol(), Symbol(), Symbol()];

class SoundManager {
  [AUDIOCONTEXT] = new (AudioContext || webkitAudioContext) ();
  [SOURCES] = { sound: null, music: null };
  [CURRENT_MUSIC] = null;
  [REFERENCES] = { sound: new WeakMap(), music: new WeakMap() };

  constructor(sound, music) {
    this[SOURCES] = { sound, music };
  }

  sound(name) {
    const t = this[SOURCES].sound[name];
    if(t instanceof Sound) {
      return t;
    }
    throw `Sound ${name} was not loaded in this room`;
  }

  music(name) {
    if(this[CURRENT_MUSIC] && this[CURRENT_MUSIC].name === name) { return; }
    if(this[CURRENT_MUSIC]) { this[CURRENT_MUSIC].stop(); }
    if(!(this[SOURCES].music[name] instanceof Sound)) { throw `Music ${name} was not loaded in this room`; }
    this[CURRENT_MUSIC] = this[SOURCES].music[name].play();
    this[CURRENT_MUSIC].then(() => {
      this[CURRENT_MUSIC] = null;
      this.music(name);
    }, () => {});
  }

  stopMusic() {
    this[CURRENT_MUSIC] && this[CURRENT_MUSIC].stop();
    this[CURRENT_MUSIC] = null;
  }

  [LOAD](type, name) {
    let sound = this[SOURCES][type][name];
    if(!sound) {
      throw `No ${type} called ${name} exists`;
    }
    if(typeof sound === 'string') {
      this[SOURCES][type][name] = new Sound(this[AUDIOCONTEXT], sound, name);
    }
    this[REFERENCES][type].set(
      this[SOURCES][type][name],
      (this[REFERENCES][type].get(this[SOURCES][type][name]) || 0) + 1
    );
    return this[SOURCES][type][name].loaded;
  }

  [PURGE](type, name) {
    if(this[SOURCES][type][name] instanceof Sound) {
      this[REFERENCES][type].set(
        this[SOURCES][type][name],
        this[REFERENCES][type].get(this[SOURCES][type][name]) - 1
      );
      if(!this[REFERENCES][type].get(this[SOURCES][type][name])) {
        this[SOURCES][type][name] = this[SOURCES][type][name].url;
      }
    }
  }

  loadSound(names) {
    names = new Set(names);
    const pr = [];
    for(let name of names) {
      pr.push(this[LOAD]('sound', name));
    }
    return Promise.all(pr);
  }

  purgeSound(names = []) {
    for(let name of names) {
      this[PURGE]('sound', name);
    }
  }

  loadMusic(names) {
    names = new Set(names);
    const pr = [];
    for(let name of names) {
      pr.push(this[LOAD]('music', name));
    }
    return Promise.all(pr);
  }

  purgeMusic(names = []) {
    for(let name of names) {
      this[PURGE]('music', name);
      if(this[CURRENT_MUSIC] && this[CURRENT_MUSIC].name === name && !(this[SOURCES].music[name] instanceof Sound)) {
        this.stopMusic();
      }
    }
  }
}

export default SoundManager;

File: c++/src/engine/sound-manager.cc
#include "sound-manager.h"
#include "sound.h"

namespace Game {
    SoundManager::SoundManager(const std::map<std::string, std::string> & sound_sources, const std::map<std::string, std::string> & music_sources)
        : _s_sources{ sound_sources }, _m_sources{ music_sources } {}

    Sound & SoundManager::sound(const std::string & name) {
        return *_sounds.at(name);
    }

    void SoundManager::music(const std::string & name) {
        if(_current_music == name) { return; }
        if(_current_music != "") {
            _music[_current_music]->stop();
        }
        _current_music = name;
        _music[_current_music]->play();
    }

    void SoundManager::stop_music() {
        if(_current_music != "") {
            _music[_current_music]->stop();
        }
        _current_music = "";
    }

    void SoundManager::load_music(const std::vector<std::string> & names) {
        for(auto & name : names) {
            if(!_music[name]) {
                _music[name] = std::make_unique<Music>(_m_sources[name]);
            }
            ++_m_references[name];
        }
    }
    void SoundManager::load_sound(const std::vector<std::string> & names) {
        for(auto & name : names) {
            if(!_sounds[name]) {
                _sounds[name] = std::make_unique<Sound>(_s_sources[name]);
            }
            ++_s_references[name];
        }
    }
    void SoundManager::purge_music(const std::vector<std::string> & names) {
        for(auto & name : names) {
            if(_music[name]) {
                if(!--_m_references[name]) {
                    if(name == _current_music) {
                        stop_music();
                    }
                    _music.erase(name);
                }
            }
        }
    }
    void SoundManager::purge_sound(const std::vector<std::string> & names) {
        for(auto & name : names) {
            if(_sounds[name]) {
                if(!--_s_references[name]) {
                    _sounds.erase(name);
                }
            }
        }
    }
}
